src/lib/x509/x509_obj.cpp

```diff
@@ -181,6 +220,49 @@ bool X509_Object::check_signature(const Public_Key& pub_key) const
       Signature_Format format =
          (pub_key.message_parts() >= 2) ? DER_SEQUENCE : IEEE_1363;
 
+      if(padding == "EMSA4")
+         {
+         // "MUST contain RSASSA-PSS-params"
+         if(signature_algorithm().parameters.empty())
+            {
+            return false;
+            }
+
+         Pss_params pss_parameter = decode_pss_params(signature_algorithm().parameters);
+
+         // hash_algo must be SHA1, SHA2-224, SHA2-256, SHA2-384 or SHA2-512
+         std::string hash_algo = OIDS::lookup(pss_parameter.hash_algo.oid);
+         if(hash_algo != "SHA-160" && hash_algo != "SHA-224" && hash_algo != "SHA-256" && hash_algo != "SHA-384"
+               && hash_algo != "SHA-512")
+            {
+            return false;
+            }
+
+         std::string mgf_algo = OIDS::lookup(pss_parameter.mask_gen_algo.oid);
+         if(mgf_algo != "MGF1")
+            {
+            return false;
+            }
+
+         // For MGF1, it is strongly RECOMMENDED that the underlying hash function be the same as the one identified by hashAlgorithm
+         // Must be SHA1, SHA2-224, SHA2-256, SHA2-384 or SHA2-512
+         std::string mgf_hash_algo = OIDS::lookup(pss_parameter.mask_gen_hash.oid);
```

Comment:
Could this just compare the OIDs directly (mask_gen_hash.oid == hash_algo.oid)? Not a big deal but seems simpler than decoding twice and comparing the strings. It would break in the case that we supported more than one OID for a hash, and someone mixed them. But I'd be inclined to reject that out of hand on principle.

Reply:
Removed the lookup and now compare the OIDs directly. Not sure if this is what you meant.